<h1>Исследование надежности заемщиков<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления.-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления.</a></span><ul class="toc-item"><li><span><a href="#Проверим-на-аномалии-столбец-days_employed." data-toc-modified-id="Проверим-на-аномалии-столбец-days_employed.-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Проверим на аномалии столбец <code>days_employed</code>.</a></span></li><li><span><a href="#Проверим-на-аномалии-столбец-children." data-toc-modified-id="Проверим-на-аномалии-столбец-children.-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Проверим на аномалии столбец <code>children</code>.</a></span></li></ul></li><li><span><a href="#Изменение-типов-данных." data-toc-modified-id="Изменение-типов-данных.-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Изменение типов данных.</a></span></li><li><span><a href="#Удаление-дубликатов." data-toc-modified-id="Удаление-дубликатов.-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Удаление дубликатов.</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Категоризация-дохода." data-toc-modified-id="Категоризация-дохода.-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация дохода.</a></span></li><li><span><a href="#Категоризация-целей-кредита." data-toc-modified-id="Категоризация-целей-кредита.-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Категоризация целей кредита.</a></span></li></ul></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-количеством-детей-заемщика-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-количеством-детей-заемщика-и-возвратом-кредита-в-срок?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Есть ли зависимость между количеством детей заемщика и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-заемщика-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-заемщика-и-возвратом-кредита-в-срок?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением заемщика и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-заемщика-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-заемщика-и-возвратом-кредита-в-срок?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода заемщика и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

# Исследование надежности заемщиков.

 На данных кредитного отдела банка разберемся, влияют ли следующие параметры на факт погашения заемщиком кредита в срок:
 - семейное положение, 
 - количество детей заемщика,
 - уровень дохода, 
 - цель кредитования. 
 
 Результаты исследования необходимы для построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Цель исследования** — получить ответы на следующие вопросы:
1. Есть ли зависимость между количеством детей заемщика и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением заемщика и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода заемщика и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

**Ход исследования**

Входные данные от банка — статистика о платёжеспособности клиентов - получим из файла `/datasets/data.csv`. О качестве данных ничего не известно, поэтому перед проверкой гипотез проведем обзор и предобработку данных. 

Таким образом, исследование пройдёт в четыре этапа:
 1. Обзор данных.
 2. Предобработка данных.
 3. Ответы на вопросы.
 4. Общий вывод.



##  Обзор данных
Составим первое представление о данных.

Основной инструмент аналитика — `pandas`. Импортируем эту библиотеку.

In [1]:
import pandas as pd

Прочитаем файл `data.csv` из пути `/Users/sofyabelaya/Downloads/data.csv` и сохраним его в переменной `data`:

In [2]:
data = pd.read_csv('/Users/sofyabelaya/Downloads/data.csv') 

Выведем на экран первые десять строк таблицы:

In [3]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Получим сводную информацию о таблице:

In [4]:
data.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Итак, в таблице 12 столбцов и 21525 строк. В столбцах используются три типа данных — `object`, `int64` и `float64`.

Согласно документации к данным:
* `children` — количество детей в семье;
* `days_employed` — общий трудовой стаж в днях;  
* `dob_years` — возраст клиента в годах;
* `education` — уровень образования клиента;
* `education_id` — идентификатор уровня образования;
* `family_status` — семейное положение;
* `family_status_id` — идентификатор семейного положения;
* `gender` — пол клиента;
* `income_type` — тип занятости;  
* `debt` —  имел ли задолженность по возврату кредитов;
* `total_income` — ежемесячный доход;
* `purpose` — цель получения кредита.

Видны следующие недочеты во входных данных: 
1. Пропущенные значения в столбцах `days_employed` и `total_income`, так как количество значений в них меньше остальных.
2. Артефакты в столбце `days_employed` - отрицательное количество дней трудового стажа.
3. Некорректный тип данных в столбце `total_income` - вещественный вместо целочисленного.
4. Неявные дубликаты в столбце `education` - значения записаны в разных регистрах.

**Выводы**

В каждой строке таблицы — полноценные данные о потенциальном заемщике: пол и возраст, семейное положение и количество детей, образование, тип занятости и стаж работы, уровень дохода и наличие задолженности по непогашеным кредитам, а также непосредственная цель кредитования. 

Предварительно можно утверждать, что, данных достаточно для получения ответов на поставленные заказчиков вопросы. Но встречаются пропуски, артефакты и неявные дубликаты  в данных.

Чтобы двигаться дальше, нужно устранить проблемы в данных.

##  Предобработка данных
Заполним пропуски и проверим данные на аномалии и исправим их. Изменим типы данных и удалим дубликаты.

### Заполнение пропусков
Сначала посчитаем, сколько в таблице пропущенных значений.

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Одинаковое количество пропусков обнаружено в столбцах `days_employed` и `total_income`. Пропущенные значения в столбце `total_income` существенны для нашего исследования. Поэтому попробуем их проанализировать и понять причину возникновения. 

Для этого:
1. проверим, какую долю составляют пропущенные значения в столбцах с пропусками;
2. проверим значения в столбцах на артефакты (аномалии);
3. проверим наличие выдающихся значений;
4. подберем наиболее подходящий вариант для заполнения пропусков.

Для начала посмотрим, пересекаются ли пропуски в обих столбцах. 

In [6]:
data[data.isna().any(axis=1)]

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

Проверим, какую долю составляют пропущенные значения в каждом из столбцов с пропусками. Поскольку количество пропусков в обоих столбцах одинаковое, то и доля пропусков также будет одинаковой. 


Посчитаем долю пропусков для столбца `total_income`.

In [7]:
'Доля пропусков: {:.0%}'.format(data['days_employed'].isna().mean())

'Доля пропусков: 10%'

Одной командой получим ответ сразу на два вопроса:
1. есть ли аномалии в виде отрицательных значений в данном столбце?
2. большой ли разброс значений?

In [8]:
data['total_income'].describe()

count    1.935100e+04
mean     1.674223e+05
std      1.029716e+05
min      2.066726e+04
25%      1.030532e+05
50%      1.450179e+05
75%      2.034351e+05
max      2.265604e+06
Name: total_income, dtype: float64

В столбце `total_income` аномалии не обнаружены. Но минимальное и максимальное значения в столбце отличаются на два порядка!

**Вывод**

Обнаруженные пропуски не похожи на простое нежелание респондента предоставить сведения, так как данные о ежемесячном доходе являются весьма важными сведениями для принятия решения банком о выдаче кредита. В обоих столбцах одинаковое количество пропусков и данные пропущены в одних и тех же строках, поэтому причинами пропусков могли стать:
* технический сбой при выгрузке данных;
* человеческий фактор.

Доля пропусков в столбцах - 10%. Процент существенный и на практике было бы правильно установить причину пропусков у заказчика и корректно восстановить данные. Но в учебном проекте такой возможности нет. 

Возможные решения:
1. пропущенные значения в столбце `total_income` - это количественные переменные и пропуски в них заполним характерными значениями. Наличие выдающихся значений (выбросов), говорит нам о том, что распределение признака не является нормальным, а значит заполнять пропуски средним арифметическим было бы некорректно, в то время как медиана к выбросам более устойчива. Так как в рамках учебной программы мы не обладаем знаниями более сложных подходов к обработке пропусков, заполним их медианой.
2. значения в столбце `days_employed` будут разобраны ниже.


Найдем медиану и заменим ей все пустые значения в столбце `total_income`.

In [9]:
total_income_median = data['total_income'].median()

In [10]:
data['total_income'] = data['total_income'].fillna(value=total_income_median)

Проверим все ли пропуски были заменены. 

In [11]:
data['total_income'].isna().sum()

0

### Проверка данных на аномалии и исправления.

#### Проверим на аномалии столбец `days_employed`.

В столбце `days_employed` обнаружены не только пропуски, но и видны аномалии в виде отрицательных значений в трудовом стаже. Посмотрим основные статистические данные об этом столбце.

In [12]:
data['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

Посчитаем долю отрицательных значений от общего числа заполненных.

In [13]:
negative_days = data[data['days_employed'] < 0]

In [14]:
'Доля отрицательных значений: {:.0%}'.format(negative_days['days_employed'].count() / data['days_employed'].count())


'Доля отрицательных значений: 82%'

Итак, столбец `days_employed` содержит следующие проблемы данных:
1. пропуски;
2. аномалии - отрицательные значения;
3. некорректный тип данных: вещественный вместо целочисленного;
4. некорректные значения: минимальный стаж - 18388 дней (50 лет), максимальный стаж - 401755 дней (1100 лет)

В рамках учебного проекта установить у заказчика причину возникновения некорректных данных невозможно. Поэтому можно только  предположить, что ошибки были допущены вследствие:
* человеческого фактора:
     - возможно, в трудовом стаже стояло тире перед значением, которое при выгрузке стало минусом,
     - возможно, раньше данные хранились в другом формате и при переносе в новую базу данных это не учли,
     - возможно, значения трудового стажа были указаны в трудочасах, а не трудоднях.
* технический сбой при выгрузке данных.

Поскольку данные этого столбца не влияют на наше дальнейшее исследование, то пока уберем только минус перед значениями.


Изменим значения в столбце `days_employed` на положительные. 

In [15]:
data['days_employed'] = data['days_employed'].abs()

Проверим все ли значения стали положительными.

In [16]:
data[data['days_employed'] < 0]['days_employed'].count()

0

#### Проверим на аномалии столбец `children`.

In [17]:
data['children'].sort_values().unique()

array([-1,  0,  1,  2,  3,  4,  5, 20])

В столбце `children` проявились неожиданные аномалии в виде отрицательного значения и выдающегося - 20 детей. 

Посмотрим статистику по количеству заемщиков с разным количеством детей.

In [18]:
data['children'].value_counts() 

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

По данным у 76 заемщиков в семье 20 детей. Выведем первые 10 строк на экран с данными о заемщиках с 20 детьми и проанализируем этот выброс.

In [19]:
data[data['children'] == 20].head(10)

children  days_employed  dob_years education  education_id  \
606         20     880.221113         21   среднее             1   
720         20     855.595512         44   среднее             1   
1074        20    3310.411598         56   среднее             1   
2510        20    2714.161249         59    высшее             0   
2941        20    2161.591519          0   среднее             1   
3302        20            NaN         35   среднее             1   
3396        20            NaN         56    высшее             0   
3671        20     913.161503         23   среднее             1   
3697        20    2907.910616         40   среднее             1   
3735        20     805.044438         26    высшее             0   

              family_status  family_status_id gender  income_type  debt  \
606         женат / замужем                 0      M    компаньон     0   
720         женат / замужем                 0      F    компаньон     0   
1074        женат / замужем                 0      F    сотрудник     1   
2510         вдовец / вдова                 2      F    сотрудник     0   
2941        женат / замужем                 0      F    сотрудник     0   
3302  Не женат / не замужем                 4      F  госслужащий     0   
3396        женат / замужем                 0      F    компаньон     0   
3671  Не женат / не замужем                 4      F    сотрудник     0   
3697       гражданский брак                 1      M    сотрудник     0   
3735  Не женат / не замужем                 4      M    сотрудник     0   

       total_income                                purpose  
606   145334.865002                          покупка жилья  
720   112998.738649                   покупка недвижимости  
1074  229518.537004                  получение образования  
2510  264474.835577  операции с коммерческой недвижимостью  
2941  199739.941398                  на покупку автомобиля  
3302  145017.937533                 профильное образование  
3396  145017.937533                     высшее образование  
3671  101255.492076     на покупку подержанного автомобиля  
3697  115380.694664     на покупку подержанного автомобиля  
3735  137200.646181                           ремонт жилью

Посмотрим какое количество респондентов младце 40 лет имеет 20 детей.

In [20]:
dob_under40 = data[data['children'] == 20]
dob_under40 = dob_under40[dob_under40['dob_years'] < 40]['dob_years'].count()
dob_under40

32

**Вывод**

Очевидно, что у заемщика в возрасте от 20 до 40 лет никак не может быть такое количество детей. Большой процент таких заемщиков дает основание предполагать, что в данные ошибочно был приписан ноль. К тому же в России зарегистрирована только одна семья с таким количеством детей - Семья Шишкиных из Воронежской области. 
Предположительно и минус в значениях стал следствием человеческой ошибки или сбоя выгрузки данных (возможно это было тире).

Исходя из полученной информации данные откорректируем следующим образом:
- значение '20'заменим на '2',
- значение '-1' заменим на '1'.

In [21]:
data['children'] = data['children'].replace({20 : 2, -1 : 1})

Проверим остались ли аномалии в столбце.

In [22]:
data['children'].value_counts() 

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

### Изменение типов данных.

В столбцe `total_income` вещественный тип данных заменим на целочисленный, так как:
 - доходы обычно округляются до целых чисел, дробная часть малоинтересна,
 - это позволит оптимизировать размер датасета.

In [23]:
data['total_income'] = data['total_income'].astype('int')

Проверим, все ли значения приведены к единому типу данных `int` и как этот столбец выглядит после измений, выведем первые 5 строк.

In [24]:
data['total_income'].dtypes

dtype('int64')

In [25]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

### Удаление дубликатов.

Проверим данные на явные и неявные дубликаты.

1. Найдем общее количество явных дубликатов в датасете, удалим их и обновим индексацию.

In [26]:
data.duplicated().sum()

54

In [27]:
data = data.drop_duplicates().reset_index(drop=True)

Проверим, все ли дубликаты удалены. В датасете было 21525 строк. После удаления 54 строк должно остаться 21471 строка.

In [28]:
data.shape

(21471, 12)

2. Найдем неявные дубликаты. 

Воспользуемся сочетанием методов `sort_values` и `unique` для нахождения неявных дубликатов во всех столбцах. В исследовании оставим только те столбцы, в которых такие дубликаты обнаружены.

In [29]:
data['education'].sort_values().unique()

array(['ВЫСШЕЕ', 'Высшее', 'НАЧАЛЬНОЕ', 'НЕОКОНЧЕННОЕ ВЫСШЕЕ',
       'Начальное', 'Неоконченное высшее', 'СРЕДНЕЕ', 'Среднее',
       'УЧЕНАЯ СТЕПЕНЬ', 'Ученая степень', 'высшее', 'начальное',
       'неоконченное высшее', 'среднее', 'ученая степень'], dtype=object)

In [30]:
data['gender'].sort_values().unique()

array(['F', 'M', 'XNA'], dtype=object)

Решим проблему с обнаруженными неявными дубликатами в данных.

* В столбце `education` приведем все значения к нижнему регистру.

In [31]:
data['education'] = data['education'].str.lower()

Проверим результат наших изменений.

In [32]:
data['education'].sort_values().unique()

array(['высшее', 'начальное', 'неоконченное высшее', 'среднее',
       'ученая степень'], dtype=object)

* В столбце `gender` была обнаружена одна строка с указанием пола 'XNA'. Поскольку пол не влияет на наше исследование, а по другим характеристикам достоверно установить пол не представляется возможным, оставим эту ячейку без изменений.

In [33]:
data[data['gender'] == 'XNA']

children  days_employed  dob_years            education  education_id  \
10690         0    2358.600502         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10690  гражданский брак                 1    XNA   компаньон     0   

       total_income               purpose  
10690        203905  покупка недвижимости

### Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.


Чтобы разгрузить исходный датафрейм и облегчить его визуальное восприятие и скорость обработки создадим два новых датафрейма со столбцами, содержащими строковые значения:
* `education_id` и `education`
* `family_status_id` и `family_status`

Получив отдельные датасеты с идентификаторами, удалим столбы `education` и  `family_status` из исходной датафрейма.


Создадим новые датасеты с идентификаторами.

In [34]:
data_education_id = data[['education', 'education_id']]

In [35]:
data_family_status_id = data[['family_status_id', 'family_status']]

Удалим дубликаты и получим два словаря-идентификатора:
1. уровень образования,
2. семейный статус.

In [36]:
data_education_id = data_education_id.drop_duplicates().reset_index(drop=True)

In [37]:
data_family_status_id = data_family_status_id.drop_duplicates().reset_index(drop=True)

Выведем словари-идентификаторы на экран.  <a id='soffee'></a>

In [38]:
data_education_id

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [39]:
data_family_status_id

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

Удалим столбы `education` и  `family_status` из исходного датафрейма `data`.

In [40]:
data = data.drop(['family_status', 'education'], axis=1)

Выведем первые 5 строк обновленного датафрейма.

In [41]:
data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0  340266.072047         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

### Категоризация дохода.  
<a id='soffee1'></a>

Для ответа на вопрос о влиянии уровня дохода заемщика на возврат кредита необходимо создать новый столбец `total_income_category` и сформировать несколько категорий дохода.

Напишем функцию для создания категорий.

In [42]:
def income_category(income):
    """
    Функция принимает значение дохода заемщика и сортирует его по категориям. 
    """
    if 0 <= income <= 30000:
        return 'E'
    elif 30001 <= income <= 50000:
        return 'D'
    elif 50001 <= income <= 200000:
        return 'C'
    elif 200001 <= income <= 1000000:
        return 'B'
    elif 1000001 <= income <= 10000000:
        return 'A'
    else:
        return 'Внимание!'

Проверим работу функции.

In [43]:
income_category(25000)

'E'

In [44]:
income_category(44000)

'D'

In [45]:
income_category(50001)

'C'

In [46]:
income_category(250000)

'B'

In [47]:
income_category(1500000)

'A'

In [48]:
income_category(-30000)

'Внимание!'

Функция работает корректно.

Добавим к исходному датафрейму `data` новый столбец `total_income_category`, где сохраним результаты применения нашей функции `income_category`.

In [49]:
data['total_income_category'] = data['total_income'].apply(income_category)

Проверим корректность применения функции к датафрейму путем подсчета категорий и выборочной проверки среза по двум категориям.


In [50]:
data['total_income_category'].value_counts() 

C    16033
B     5041
D      350
A       25
E       22
Name: total_income_category, dtype: int64

In [51]:
data['total_income_category'].count() 

21471

In [52]:
data[data['total_income_category'] == 'E'].head()

children  days_employed  dob_years  education_id  family_status_id  \
934          0  348253.361682         69             1                 0   
1184         0  401575.196728         55             1                 1   
1598         0  359726.104207         68             1                 1   
4056         0  386389.497407         64             1                 0   
8268         0     615.774147         46             1                 0   

     gender income_type  debt  total_income  \
934       F   пенсионер     0         29426   
1184      M   пенсионер     0         29749   
1598      M   пенсионер     0         21695   
4056      F   пенсионер     1         29154   
8268      F   сотрудник     0         26533   

                                purpose total_income_category  
934                        недвижимость                     E  
1184              на проведение свадьбы                     E  
1598              на проведение свадьбы                     E  
4056         строительство недвижимости                     E  
8268  покупка коммерческой недвижимости                     E

In [53]:
data[data['total_income_category'] == 'A'].head()

children  days_employed  dob_years  education_id  family_status_id  \
1590         1    1283.436610         53             0                 0   
2224         1    4614.851849         42             0                 0   
4128         0    7992.968249         52             0                 0   
4413         0     542.321356         45             1                 1   
7443         0    8626.777234         39             1                 3   

     gender income_type  debt  total_income  \
1590      F   компаньон     0       1089120   
2224      M   сотрудник     0       1278622   
4128      F   компаньон     0       1004476   
4413      M   компаньон     1       1030899   
7443      F   сотрудник     0       1240165   

                                    purpose total_income_category  
1590                  на покупку автомобиля                     A  
2224                          покупка жилья                     A  
4128  операции с коммерческой недвижимостью                     A  
4413                 профильное образование                     A  
7443                   покупка недвижимости                     A

### Категоризация целей кредита.

Для ответа на вопрос о влиянии цели кредита на возврат кредита необходимо создать новый столбец `purpose_category` и сформировать несколько категорий.

Для начала рассмотрим все уникальные значения в столбце `purpose`.

In [54]:
data['purpose'].sort_values().unique()

array(['автомобили', 'автомобиль', 'высшее образование',
       'дополнительное образование', 'жилье',
       'заняться высшим образованием', 'заняться образованием',
       'на покупку автомобиля', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля', 'на проведение свадьбы',
       'недвижимость', 'образование', 'операции с жильем',
       'операции с коммерческой недвижимостью',
       'операции с недвижимостью', 'операции со своей недвижимостью',
       'покупка жилой недвижимости', 'покупка жилья',
       'покупка жилья для сдачи', 'покупка жилья для семьи',
       'покупка коммерческой недвижимости', 'покупка недвижимости',
       'покупка своего жилья', 'получение высшего образования',
       'получение дополнительного образования', 'получение образования',
       'приобретение автомобиля', 'профильное образование',
       'ремонт жилью', 'свадьба', 'свой автомобиль',
       'сделка с автомобилем', 'сделка с подержанным автомобилем',
       'строительство 

На основании этих целей выявим и применим единые корни слов для создания функции.

Напишем функцию для создания категорий. <a id='soffee2'></a>

In [55]:
def purpose_category(purpose):
    """
    Функция принимает значение цели кредитования и сортирует ее по категориям. 
    """
    if 'автомобил'in purpose:
        return 'операции с автомобилем'
    elif 'жиль' in purpose or 'недвижимост' in purpose:
        return 'операции с недвижимостью'
    elif 'свадьб' in purpose:
        return 'проведение свадьбы'
    elif 'образован' in purpose:
        return 'получение образования'
    else:
        return 'Новая цель!'

Проверим работу функции.

In [56]:
purpose_category('покупка жилья для сдачи')

'операции с недвижимостью'

In [57]:
purpose_category('строительство собственной недвижимости')

'операции с недвижимостью'

In [58]:
purpose_category('сыграть свадьбу')

'проведение свадьбы'

In [59]:
purpose_category('получение высшего образования')

'получение образования'

In [60]:
purpose_category('сделка с подержанным автомобилем')

'операции с автомобилем'

In [61]:
purpose_category('покупка частного дома')

'Новая цель!'

Функция работает корректно.

Добавим к исходному датафрейму `data` новый столбец `purpose_category`, где сохраним результаты применения нашей функции `purpose_category`.

In [62]:
data['purpose_category'] = data['purpose'].apply(purpose_category)

Проверим корректность применения функции к датафрейму путем подсчета категорий и выборочной проверки среза по двум категориям.

In [63]:
data['purpose_category'].value_counts() 

операции с недвижимостью    10814
операции с автомобилем       4308
получение образования        4014
проведение свадьбы           2335
Name: purpose_category, dtype: int64

In [64]:
data[data['purpose_category'] == 'проведение свадьбы'].head()

children  days_employed  dob_years  education_id  family_status_id gender  \
4          0  340266.072047         53             1                 1      F   
8          2    6929.865299         35             0                 1      F   
12         0            NaN         65             1                 1      M   
23         0     272.981385         21             0                 1      M   
32         0    4649.910832         34             1                 1      F   

   income_type  debt  total_income                purpose  \
4    пенсионер     0        158616        сыграть свадьбу   
8    сотрудник     0         95856  на проведение свадьбы   
12   пенсионер     0        145017        сыграть свадьбу   
23   сотрудник     0        128265        сыграть свадьбу   
32   сотрудник     1        139057  на проведение свадьбы   

   total_income_category    purpose_category  
4                      C  проведение свадьбы  
8                      C  проведение свадьбы  
12                     C  проведение свадьбы  
23                     C  проведение свадьбы  
32                     C  проведение свадьбы

In [65]:
data[data['purpose_category'] == 'операции с автомобилем'].head()

children  days_employed  dob_years  education_id  family_status_id gender  \
1          1    4024.803754         36             1                 0      F   
13         0    1846.641941         54             2                 0      F   
18         0  400281.136913         53             1                 2      F   
19         0   10038.818549         48             1                 3      F   
21         1     253.685166         33             1                 1      F   

   income_type  debt  total_income                             purpose  \
1    сотрудник     0        112080             приобретение автомобиля   
13   сотрудник     0        130458             приобретение автомобиля   
18   пенсионер     0         56823  на покупку подержанного автомобиля   
19   сотрудник     0        242831        на покупку своего автомобиля   
21   сотрудник     0        131812             приобретение автомобиля   

   total_income_category        purpose_category  
1                      C  операции с автомобилем  
13                     C  операции с автомобилем  
18                     C  операции с автомобилем  
19                     B  операции с автомобилем  
21                     C  операции с автомобилем

**Вывод** 

В процессе предобработки данных были обнаружены и исправлены следующие проблемы:

* пропущенные значения в каждлом отдельном случае заменены на релевантные,
* изменены типы данных в столбцах на наиболее уместные,
* удалены дубликаты,
* исправлены выявленные аномалии в данных,
* сформированы словари-идентификаторы для визуального удобства восприятия,
* произведена необходимая категоризация данных.

    Благодаря проделанной предварительной обработке данных исходный датафрейм был исправлен и значительно разгружен, а также облегчено его визуальное восприятие и скорость обработки.

Теперь можно перейти к проверке гипотез.

## Ответы на вопросы.

### Есть ли зависимость между количеством детей заемщика и возвратом кредита в срок?

Для ответа на этот вопрос рассчитаем долю дефолта по каждой группе.

In [66]:
data.groupby( ['children'] ).agg({'debt': 'mean'})

debt
children          
0         0.075353
1         0.091639
2         0.094925
3         0.081818
4         0.097561
5         0.000000

In [67]:
# Код ревьюера

data.groupby('children')['debt'].mean()

children
0    0.075353
1    0.091639
2    0.094925
3    0.081818
4    0.097561
5    0.000000
Name: debt, dtype: float64

**Вывод**

Из данных видно, что существенной зависимости между количеством детей заемщика и возвратом кредита нет. В среднем в каждой категории 'количество детей заемщика' одинаковый процент должников от 7,5% до 10%. Максимальный процент невозврата кредита - 9,8% - у заемщиков с 4 детьми. Заемщиков с 5 детьми слишком мало, поэтому выборку нельзя считать репрезентативной и влияющей на результаты исследования. Поэтому минимальный процент невозврата кредита - 7,5% - у заемщиков без детей.

### Есть ли зависимость между семейным положением заемщика и возвратом кредита в срок?

Для ответа на этот вопрос оценим уровень дефолта в каждой категории с помощью группировки данных.

In [68]:
data.groupby( ['family_status_id'] ).agg({'debt': 'mean'})

debt
family_status_id          
0                 0.075421
1                 0.093202
2                 0.065693
3                 0.071130
4                 0.097509

Для удобства и быстроты понимания уровня дефолта выведем словарь-идентификатор семейного положения, полученный в [п. 2.5.](#soffee) нашего исследования.

In [69]:
data_family_status_id

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

**Вывод**

 Данные показывают, что кардинально семейное положение заемщика не влияет на возврат кредита, так как процент должников колеблется в пределах от 6,6% до 9,75%. 

Самый низкий процент задолженников во второй категории -'вдовец/вдова' - 6,6%. При этом максимальный порог невозврата кредита среди всех категорий не превышает 9,75% и является показателем в четвертой катгории - 'не женат/не замужем'.

### Есть ли зависимость между уровнем дохода заемщика и возвратом кредита в срок?

Для ответа на этот вопрос рассчитаем долю дефолта по каждой категории заемщиков. Для этого воспользуемся данными из нового столбца с категоризацией дохода total_income_category , полученный в [п.2.6.](#soffee1) нашего исследования.

In [70]:
data.groupby( ['total_income_category'] ).agg({'debt': 'mean'})

debt
total_income_category          
A                      0.080000
B                      0.070621
C                      0.084825
D                      0.060000
E                      0.090909

Для удобства и быстроты понимания значений в категориях дохода заемщика вспомним функцию для их расчета.

In [71]:
def income_category(income):
    if 0 <= income <= 30000:
        return 'E'
    elif 30001 <= income <= 50000:
        return 'D'
    elif 50001 <= income <= 200000:
        return 'C'
    elif 200001 <= income <= 1000000:
        return 'B'
    elif 1000001 <= income <= 10000000:
        return 'A'
    else:
        return 'Внимание!'

**Вывод**

Данные показывают, что принципиально доход заемщика не влияет на возврат кредита и процент невозврата кредита колеблется от 6 до 9% во всех категориях.

Однако стоит отметить, что самый большой процент невозврата - 9% у заемщиков категории 'Е' с самым низким уровнем дохода - менее 30 тысяч рублей. При этом самыми ответственными заемщиками можно считать категорю 'D' - с уровнем дохода от 30 до 50 тысяч реблей - здесь уровень задолженников не превышает 6%. 

### Как разные цели кредита влияют на его возврат в срок?

Для ответа на этот вопрос рассчитаем долю дефолта по каждой категории. Для этого воспользуемся данными из нового столбца с категоризацией цели кредита и получен новый столбец `purpose_category`, полученный в [п.2.7.](#soffee2) нашего исследования.

In [72]:
data.groupby( ['purpose_category'] ).agg({'debt': 'mean'})

debt
purpose_category                  
операции с автомобилем    0.093547
операции с недвижимостью  0.072314
получение образования     0.092177
проведение свадьбы        0.079657

**Вывод**

Данные показывают, что цель кредита особенно не влияет на возврат кредита и процент невозврата кредита колеблется от 7 до 9% во всех категориях.

Однако стоит отметить, что самый большой процент невозврата - 9% у заемщиков, получающих кредит нв 'операции с автомобилем'. При этом более ответственными заемщиками можно считать тех, кто берет кредит на 'операции с недвижимостью' - процент невозврата здесь не превышает 7,2%. 

## Общий вывод:

   В результате исследования были обнаружены пропуски, многочисленные аномальные значения, которые могли существенно исказить результаты анализа. Тем не менее благодаря проведенному полноценному анализу и корректировки полученных от заказчика данных, можно сделать следующие выводы и предложить необходимые меры для улучшения сбора данных.
   В процессе предобработки возникли сложности с данными, поэтому заказчику в первую очередь рекомендуем наладить контроль за внесением значений в единую базу данных и исправить существующие несоответствия:
   1. внести полноценные и достоверные сведения в столбец `total_income`
   2. откорректировать описание значений в столбце `days_employed` и указать корректные единицы измерения. Так как существующее описание говорит нам о невероятном трудовом стаже заемщиков от 50 до 1100 лет! Обратить внимание на отрицательные значения стажа.
   3. обратить внимание на столбец `children`, имеющий как отрицательные, так и аномально большие значения.
   
   
   Целью исследования было понять влияют ли следующие параметры на факт погашения заемщиком кредита в срок:
* семейное положение,
* количество детей заемщика,
* уровень дохода,
* цель кредитования.

   Проведенный анализ данных помог охарактеризовать основного потребителя услуги кредитования в банке:
   
   1. самая большая категория заемщиков не имеет детей. На их долю приходится 61% всех заемщиков, исследуемых в нашей работе.
   2. самая большая категория заемщиков - '0' - 'женат / замужем. На ее долю приходится 53% всех заемщиков, исследуемых в нашей работе.
   3. самая большая категория заемщиков - 'C' - с уровнем джохода от 50 до 200 тысяч рублей. На ее долю приходится 68% респондентов. 
   4. основная цель кредитования у заемщиков - 'операции с недвижимостью'. С этой целью за кредитом обратились 47% заемщиков. 
  
При этом были получены основыне характеристики стабильного и дефолтного заемщика:

* МИНИМАЛЬНЫЕ риски по невозврату кредита лежат на следующем заемщике:
   - вдовец / вдова,
   - не имеет детей,
   - имеет доход от 30 до 50 тысяч рублей в месяц,
   - цель кредита - 'операции с недвижимостью'.
   
  
* МАКСИМАЛЬНЫЕ риски по невозврату кредита лежат на следующем заемщике:
   - не женат / не замужем,
   - с детьми (чем больше детей, тем больше вероятность невозврата кредита),
   - имеет доход менее 30 тысяч рублей в месяц,
   - цель кредита - 'операции с автомобилем'.
   
   В целом можно утверждать, что кардинальной зависимости между количеством детей у заемщика, ровно как и его семейным положением - нет.
В расширении продуктов кредитования банку стоит опираться в первую очередь на стабильного заемщика в лице семейного человека  с хорошим стабильным достатком от 50 до 200 тыс. рублей, ставящим целью приобретение или расширений существующей жилплощади. 
В свою очередь стоит помнить о максимальных рисках с молодыми людьми, без стабильного семейного положения, с мальнькими и, возможно, не постояннными заработками, целью кредита которых является покупка автомобиля. Возможно стоит пересмотреть условия выдачи кредита, прибегнув к более тщательной проверки платежеспособности будущего заемщика. 
   